tracing-subscriber/src/filter/layer_filters.rs

```diff
@@ -373,50 +731,410 @@ where
 // === impl FilterId ===
 
 impl FilterId {
+    const DISABLED: FilterId = FilterId(u64::MAX);
+
     pub(crate) fn new(id: u8) -> Self {
         assert!(id < 64, "filter IDs may not be greater than 64");
-        Self(id)
+        Self(1 << id as usize)
+    }
+
+    pub(crate) fn and(self, FilterId(other): Self) -> Self {
+        if self.0 == Self::DISABLED.0 {
+            return Self(other);
+        }
+
+        Self(self.0 | other)
+    }
+}
+
+impl fmt::Debug for FilterId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        // don't print a giant set of the numbers 0..63 if the filter ID is disabled.
+        if self.0 == Self::DISABLED.0 {
+            return f
+                .debug_tuple("FilterId")
+                .field(&format_args!("DISABLED"))
+                .finish();
+        }
+
+        if f.alternate() {
+            f.debug_struct("FilterId")
+                .field("ids", &format_args!("{:?}", FmtBitset(self.0)))
+                .field("bits", &format_args!("{:b}", self.0))
+                .finish()
+        } else {
+            f.debug_tuple("FilterId").field(&FmtBitset(self.0)).finish()
+        }
+    }
+}
+
+impl fmt::Binary for FilterId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("FilterId")
+            .field(&format_args!("{:b}", self.0))
+            .finish()
     }
 }
 
 // === impl FilterMap ===
 
 impl FilterMap {
-    pub(crate) fn set(self, FilterId(idx): FilterId, enabled: bool) -> Self {
-        debug_assert!(idx < 64 || idx == 255);
-        if idx >= 64 {
+    pub(crate) fn set(self, FilterId(mask): FilterId, enabled: bool) -> Self {
+        if mask == u64::MAX {
             return self;
         }
 
         if enabled {
             Self {
-                bits: self.bits & !(1 << idx),
+                bits: self.bits & (!mask),
             }
         } else {
             Self {
-                bits: self.bits | (1 << idx),
+                bits: self.bits | mask,
```

Comment:
Why this change, outta curiosity?

Reply:
ah, this is something i wanted to add additional comments explaining. essentially, this is in order to make filtering the span traversal work when you have a situation like this:

````rust
Filtered {
    filter1,
    Layered {
        layer1,
        Filtered {
             filter2, 
             layer2,
        },
}
````
and you want `layer2` to be effected by both `filter1` _and_ `filter2`. With the previous implementation, this worked fine when filtering `on_event`/`new_span`. However, it _doesn't_ work when filtering span lookups and traversals (e.g. `scope`). This is because the `Context` passed to `layer2` would set its filter ID to the filter ID of `filter2`, and would skip spans that were disabled by `filter2`. However, what if a span was disabled by `filter1`? We wouldn't see it in `new_span`, but we _would_ see it in lookups and traversals...which we don't want.

Now, when a `Filtered` layer adds its ID to a `Context`, it _combines_ it with any previous filter ID that the context had, rather than replacing it. That way, `layer2`'s context will check if a span was disabled by `filter1` _or_ `filter2`. The way I did this is by changing the `FilterId`s from being a number that we shift a 1 over by to get a mask to just _being_ the mask, so we can combine them with a bitwise-OR.

For example, if we assume that 
- `filter1` has the filter id 1 (`0b0000_0001`) 
- `filter2` has the filter id 2 (`0b0000_0010`)

A span that gets disabled by filter 1 would have the `FilterMap` with bits `0b0000_0001`.

If we did the previous approach, and have the filter ID be the number of bits to shift + 1, when `layer2`'s context checked if it enabled the span, it would make the mask `0b0000_0010` (`1 << 1`). That bit would not be set in the filtermap, so it would see that it _didn't_ disable the span. Which is *true*, it just doesn't reflect the tree-like shape of the actual subscriber.

By having the IDs be masks instead of shifts, though, when the `Filtered` with `filter2` gets the context with `filter1`'s filter ID, instead of replacing it, it ors them together: `0b0000_0001 | 0b0000_0010` -> `0b0000_0011`. We then test if the span was enabled by seeing if _any_ bits in the mask are `1`: 
```rust
filtermap & mask != 0
0b0000_0001 & 0b0000_0011 != 0
0b0000_0001 != 0
true
```

does that make sense?

Reply:
the other thing is that if we pre-compute the mask, we avoid a single bitshift operation when testing if something is enabled in a filtermap. bitshifts are _extremely_ cheap in most CPU architectures, so this is _probably_ not a significant optimization on its own, but it's a nice side benefit of something which was necessary for other reasons.

Reply:
This does make sense, thank you! Copying and pasting what you _exactly_ wrote into the (private) doc comments would be wildly helpful, I think.

Additionally, do you think it makes sense to have tests that compare the results of filtering with a bit-based `FilterMap` against a (slower, but whose behavior is _not_ novel) `HashSet`?

Reply:
> This does make sense, thank you! Copying and pasting what you _exactly_ wrote into the (private) doc comments would be wildly helpful, I think.

done in ee1c8d2c04467f978f7bc23e125342003a0b9033

> Additionally, do you think it makes sense to have tests that compare the results of filtering with a bit-based `FilterMap` against a (slower, but whose behavior is _not_ novel) `HashSet`?

hmm...i'm not convinced this is really necessary, as the bitmap implementation is quite simple, and it's not really particularly "novel". we could _do_ it, but I dunno if it's really worth the effort.